File: TypeScript/ExampleTests/common/DiscountingTyped.ts
import { DiscountingString } from "./DiscountingString.js";
import * as _json from "./json.js";

export class DiscountingTyped {
  totalPrice: string;
  discount: string;
  notes: string;

  constructor(totalPrice: string, discount: string, notes: string) {
    this.totalPrice = totalPrice;
    this.discount = discount;
    this.notes = notes;
  }

  static fromStringObj(s: DiscountingString): DiscountingTyped {
    return new DiscountingTyped(
      s.totalPrice,
      s.discount,
      s.notes
    );
  }

  toJsonValue(): Record<string, unknown> {
    return {
      totalPrice: this.totalPrice,
      discount: this.discount,
      notes: this.notes,
    };
  }

  toJSON(): string { return _json.stringify(this.toJsonValue()); }

  static fromJsonValue(m: unknown): DiscountingTyped {
    return new DiscountingTyped(
      _json.asString(_json.requireField(m, "totalPrice"), "totalPrice"),
      _json.asString(_json.requireField(m, "discount"), "discount"),
      _json.asString(_json.requireField(m, "notes"), "notes")
    );
  }

  static fromJSON(text: string): DiscountingTyped {
    return DiscountingTyped.fromJsonValue(_json.parse(text));
  }

  static toJSONList(items: readonly DiscountingTyped[]): string {
    return _json.stringify(items.map((item) => item.toJsonValue()));
  }

  static fromJSONList(text: string): DiscountingTyped[] {
    const raw = _json.asArray(_json.parse(text), "DiscountingTyped") ?? [];
    return raw.map((e) => DiscountingTyped.fromJsonValue(e));
  }

  toString(): string {
    return `Total Price=${this.totalPrice}, Discount=${this.discount}, Notes=${this.notes}`;
  }

  equals(other: DiscountingTyped): boolean {
    return this.totalPrice === other.totalPrice
      && this.discount === other.discount
      && this.notes === other.notes;
  }
}

File: Python/ExampleTests/common/discounting_typed.py
from . import json_util as _json
from .discounting_string import DiscountingString

class DiscountingTyped:
    def __init__(self, total_price: str = '', discount: str = '', notes: str = ''):
        self.total_price = total_price
        self.discount = discount
        self.notes = notes

    @classmethod
    def from_string_obj(cls, s: DiscountingString) -> 'DiscountingTyped':
        return cls(
            s.total_price,
            s.discount,
            s.notes
        )

    def to_json_value(self) -> dict:
        return {
            'total_price': self.total_price,
            'discount': self.discount,
            'notes': self.notes,
        }

    def to_json(self) -> str:
        return _json.dumps(self.to_json_value())

    @classmethod
    def from_json_value(cls, m: dict) -> 'DiscountingTyped':
        return cls(
            _json.as_str(_json.require(m, 'total_price'), 'total_price'),
            _json.as_str(_json.require(m, 'discount'), 'discount'),
            _json.as_str(_json.require(m, 'notes'), 'notes')
        )

    @classmethod
    def from_json(cls, text: str) -> 'DiscountingTyped':
        return cls.from_json_value(_json.loads(text))

    @staticmethod
    def to_json_list(items) -> str:
        return _json.dumps([item.to_json_value() for item in items])

    @classmethod
    def from_json_list(cls, text: str) -> list:
        raw = _json.as_list(_json.loads(text), 'DiscountingTyped')
        return [cls.from_json_value(e) for e in raw]
